bdns/dns.go

```diff
@@ -283,29 +321,79 @@ func isPrivateV4(ip net.IP) bool {
 	return false
 }
 
-// LookupHost sends a DNS query to find all A records associated with the
-// provided hostname. This method assumes that the external resolver will chase
-// CNAME/DNAME aliases and return relevant A records.  It will retry requests in
-// the case of temporary network errors. It can return net package,
-// context.Canceled, and context.DeadlineExceeded errors.
-func (dnsResolver *DNSResolverImpl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
-	var addrs []net.IP
-	dnsType := dns.TypeA
-	r, err := dnsResolver.exchangeOne(ctx, hostname, dnsType, dnsResolver.aStats)
-	if err != nil {
-		return addrs, &DNSError{dnsType, hostname, err, -1}
+func isPrivateV6(ip net.IP) bool {
+	for _, net := range privateV6Networks {
+		if net.Contains(ip) {
+			return true
+		}
 	}
-	if r.Rcode != dns.RcodeSuccess {
-		return nil, &DNSError{dnsType, hostname, nil, r.Rcode}
+	return false
+}
+
+// LookupHost sends a DNS query to find all A and AAAA records associated with
+// the provided hostname. This method assumes that the external resolver will
+// chase CNAME/DNAME aliases and return relevant records.  It will retry
+// requests in the case of temporary network errors. It can return net package,
+// context.Canceled, and context.DeadlineExceeded errors, all wrapped in the
+// DNSError type.
+func (dnsResolver *DNSResolverImpl) LookupHost(ctx context.Context, hostname string) ([]net.IP, error) {
+	var recordsA, recordsAAAA []dns.RR
+	var errA, errAAAA error
+	var wg sync.WaitGroup
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		respA, err := dnsResolver.exchangeOne(ctx, hostname, dns.TypeA, dnsResolver.aStats)
+		if err != nil {
+			errA = &DNSError{dns.TypeA, hostname, err, -1}
+			return
+		}
+		if respA.Rcode != dns.RcodeSuccess {
+			errA = &DNSError{dns.TypeA, hostname, nil, respA.Rcode}
+			return
+		}
+		recordsA = respA.Answer
+	}()
+	go func() {
+		defer wg.Done()
+		if !dnsResolver.LookupIPv6 {
```

Comment:
Only spawn a goroutine if `LookupIPv6 == true`, also above split the `wg.Add` call into two and only add a second if you are going to spawn the goroutine.